File: engine/scripting/dotnet/src/RetroEngine/Interop/EntityExporter.cs
// @file $EntityExporter.cs.cs
//
// @copyright Copyright (c) 2026 Retro & Chill. All rights reserved.
// Licensed under the MIT License. See LICENSE file in the project root for full license information.
using RetroEngine.Binds;
using RetroEngine.Core.Math;
using RetroEngine.Scene;

namespace RetroEngine.Interop;

[BindExport("retro")]
internal static partial class EntityExporter
{
    public static partial int GetEntityTransformOffset();

    [return: CppType(TypeName = "retro::Entity", CppModule = "retro.runtime", UseReference = true)]
    public static partial IntPtr CreateNewEntity(in Transform transform, out EntityId id);

    public static partial void RemoveEntityFromScene(
        [CppType(TypeName = "retro::Entity", CppModule = "retro.runtime", IsConst = true, UseReference = true)]
            IntPtr entityPtr
    );
}

File: engine/runtime/src/interop/entity_exporter.cpp
/**
 * @file entity_exporter.cpp
 *
 * @copyright Copyright (c) 2026 Retro & Chill. All rights reserved.
 * Licensed under the MIT License. See LICENSE file in the project root for full license information.
 */
module retro.runtime.interop;

namespace retro::entity_exporter
{
    const Transform &get_entity_transform(const EntityID entity_id)
    {
        return Engine::instance().scene().get_entity(entity_id).value().transform();
    }

    void set_entity_transform(const EntityID entity_id, const Transform &transform)
    {
        Engine::instance().scene().get_entity(entity_id).value().set_transform(transform);
    }

    EntityID create_new_entity(const Transform &transform)
    {
        return Engine::instance().scene().create_entity(transform).id();
    }

    void remove_entity_from_scene(const EntityID entity_id)
    {
        Engine::instance().scene().destroy_entity(entity_id);
    }
} // namespace retro::entity_exporter
